RecoParticleFlow/PFClusterProducer/src/PFClusterEMEnergyCorrector.cc

```diff
@@ -116,156 +140,367 @@ void PFClusterEMEnergyCorrector::correctEnergies(const edm::Event &evt, const ed
+      for (auto i_ps = clustops.first; i_ps != clustops.second; ++i_ps) {
         edm::Ptr<reco::PFCluster> psclus(i_ps->second);
+	       
         switch( psclus->layer() ) {
         case PFLayer::PS1:
           ePS1 += psclus->energy();
           break;
         case PFLayer::PS2:
           ePS2 += psclus->energy();
-          break;
+	  break;
         default:
-          break;
+	  break;
         }
       }
     }
-    
+
     //fill array for forest evaluation
-    eval[0] = evale;
-    eval[1] = ietaix;
-    eval[2] = iphiiy;
-    if (!iseb) {
-      eval[3] = ePS1*invE;
-      eval[4] = ePS2*invE;
+    if (iseb) {
+      evalEB[0] = evale;
+      evalEB[1] = ietaix;
+      evalEB[2] = iphiiy;
+      evalEB[3] = ietamod20;
+      evalEB[4] = iphimod20;
+      evalEB[5] = reducedHits;
+    } else {
+      evalEE[0] = evale;
+      evalEE[1] = ietaix;
+      evalEE[2] = iphiiy;
+      evalEE[3] = (ePS1+ePS2)*invE;
+      evalEE[4] = reducedHits;
     }
-        
+     
     //these are the actual BDT responses
-    double rawmean = meanforest.GetResponse(eval.data());
-    double rawsigma = sigmaforest.GetResponse(eval.data());
-    
+    double rawmean = 1;
+    double rawsigma = 0;
+
+    if(iseb){
+      rawmean = meanforest.GetResponse(evalEB.data());
+      rawsigma = sigmaforest.GetResponse(evalEB.data());
+    } else {
+      rawmean = meanforest.GetResponse(evalEE.data());
+      rawsigma = sigmaforest.GetResponse(evalEE.data());
+    }
+
     //apply transformation to limited output range (matching the training)
     double mean = meanoffset + meanscale*vdt::fast_sin(rawmean);
     double sigma = sigmaoffset + sigmascale*vdt::fast_sin(rawsigma);
     
     //regression target is ln(Etrue/Eraw)
     //so corrected energy is ecor=exp(mean)*e, uncertainty is exp(mean)*eraw*sigma=ecor*sigma
-    double ecor = exp(mean)*e;
+    double ecor = iseb ? vdt::fast_exp(mean)*e : vdt::fast_exp(mean)*(e+ePS1+ePS2);
     double sigmacor = sigma*ecor;
+
+    LogDebug("PFClusterEMEnergyCorrector") << "ieta : iphi : ietamod20 : iphimod20 : size : reducedHits = "
+					   << ietaix << " " << iphiiy << " " 
+					   << ietamod20 << " " << iphimod20 << " " 
+					   << size << " " << reducedHits
+					   << "\n" << "isEB : eraw : ePS1 : ePS2 : (eps1+eps2)/raw : Flag = "
+					   << iseb << " " << evale << " " << ePS1 << " " << ePS2 << " " << (ePS1+ePS2)/evale << " " << clusFlag
+					   << "\n" << "response : correction = " 
+					   << exp(mean) << " " << ecor;
     
     cluster.setCorrectedEnergy(ecor);
     cluster.setCorrectedEnergyUncertainty(sigmacor);
-    
   }
   
 }
 
+EcalTrigTowerDetId PFClusterEMEnergyCorrector::readOutUnitOf(const EBDetId& xtalId) const{
+  return triggerTowerMap_->towerOf(xtalId);
+}
 
-
+EcalScDetId PFClusterEMEnergyCorrector::readOutUnitOf(const EEDetId& xtalId) const{
+  const EcalElectronicsId& EcalElecId = elecMap_->getElectronicsId(xtalId);
+  int iDCC= EcalElecId.dccId();
```

Comment:
Ok, I am adding the functions in the suggested package. If ECAL reco wants it in a different package, I will be happy to move it there.